content/blog/2023-12-11-circ-bitwidth-adjustment

```diff
@@ -0,0 +1,115 @@
+
+## Design
+Our optimization mainly focuses on addition and multiplication, which would typically increase the bitwidth of resulting element by 1 or double the bitwidth, respectively. We classify elements into two types
+- Elements that could be above the range and required to be returned within range: this includes program output, elements involved in bitwise operations and comparisons, operations like division or remainder, elements involved in memory operations, and so on.
+- Elements that could be above the range but are not required to be within range: This includes intermediate elements that undergoes multiplication and addition operations and does not meet any of the criterions above.
+
+The current system represents basic elements as bitvectors, and is designed to adjust their bitwidth after each arithmetic operation. Based on the classification of the elements above, our projects reduces the number of constraints in the system by greedily deciding when to adjust bitwidth for elements falls under the second category while ensuring no overflows can happen in later operations involving the elements. Or in other words, adjustments are only introduced if the next operation is going to result in an overflow.
+
+Take this short program as an example:
+```
+def main(u32 a, u32 b) -> u32:
+    u32 c = a + b
+    u32 d = a + c
+    return c * d
+```
+
+The old compiler would adjust immediately at `a+b`, `a+c` and `c * d`. By applying our optimization, the system would treat c and d as intermediate elements and only adjust at c*d, since it is producing the final program output. 
```

Comment:
I think a major part of my confusion here is that I don't yet see what "adjust" means. Does it mean that we represent `c`, `d`, and the return value of `main` using wider types than `u32`? That seems a little odd because the source code implies that, for example, `c` must fit within 32 bits. The only semantics I am aware of out there for any programming language with fixed-width types like `u32` are:

* overflow wraps around
* overflow is a dynamic error
* overflow is undefined behavior

I'm not really familiar with any frontend language that says "overflow is represented precisely, and we ignore the fact that you declared the variable as a `u32` and represent it with an arbitrary-precision BigInt anyway." Is that the semantics of the language you're working with here?

I think I have missed the point and "adjust" means something more subtle than "use more bits to represent it." Can you explain in more detail what the "adjustment" consists of?